src/main/java/com/hierynomus/smbj/share/DiskShare.java

```diff
@@ -434,17 +565,115 @@ public void setSecurityInfo(SMB2FileId fileId, Set<SecurityInformation> security
         );
     }
 
+    /***
+     * 3.2.5.19 Receiving an SMB2 OPLOCK_BREAK Notification, this handler is responsible to call acknowledgeOplockBreak if needed. Set the handler for Receiving an Oplock Break Notification.
+     * You MUST set this handler before create/open diskEntry with oplock.
+     *
+     * @param handler handler for Receiving an Oplock Break Notification and Async Create Request/Response.
+     */
+    public void setNotificationHandler(NotificationHandler handler) {
+        this.notificationHandler = handler;
+    }
+
+    /***
+     * Handler for handing the oplock break notification event from server. 3.2.5.19 Receiving an SMB2 OPLOCK_BREAK Notification.
+     *
+     * @param oplockBreakNotification received oplock break notification from server.
+     */
+    @Handler
+    @SuppressWarnings("unused")
+    private void oplockBreakNotification(final OplockBreakNotification oplockBreakNotification) {
+        try {
+
+            final SMB2FileId fileId = oplockBreakNotification.getFileId();
```

Comment:
Shouldn't we check here whether the `OplockBreakNotification.fileId` is actually handled by this `DiskShare`?

Reply:
I don't think we need to check for fileId. The client **MUST** ignore it if `fileId` not found.

Base on document,
>3.2.5.1.2 Finding the Application Request for This Response
>The client MUST locate the request for which this response was sent in reply by locating the request in
Connection.OutstandingRequests using the MessageId field of the SMB2 header. If the request is
not found, the response MUST be discarded as invalid.
If the MessageId is 0xFFFFFFFFFFFFFFFF, this is not a reply to a previous request, and the client
MUST NOT attempt to locate the request, but instead process it as follows:
If the command field in the SMB2 header is SMB2 OPLOCK_BREAK, it MUST be processed as specified
in 3.2.5.19. Otherwise, the response MUST be discarded as invalid.

>3.2.5.19.1 Receiving an Oplock Break Notification
>The client MUST locate the open in the Session.OpenTable using the FileId in the Oplock Break
Notification following the SMB2 header. If the open is **not found**, the oplock break indication **MUST be
discarded**, and **no further processing is required**.

And currently, it seems the `DiskShare` didn't record for the openedHandle. I don't think we should add it just because an unnecessary checking.

But indeed reminds me something. The `fileId` is only unique within a `session`. We should check the `oplockBreakNotification.getHeader().getSessionId()` match `DiskShare.session.getSessionId()` or not before notifying the user.

Reply:
At this moment multiple `DiskShare`s will be receiving the message from the bus. You need to check which one needs to process the message, so you _do_ need to check both the `sessionId` as well as the `fileId`.

Reply:
Just to be sure, is it sufficient to confirm the `DiskShare` is same by checking the `oplockBreakNotification.getHeader().getTreeId()` match `DiskShare.treeConnect.getTreeId()`?

Reply:
Ok, that makes that we need to use a contextual bus per connection. Because it could very well be that the same fileId exists on multiple different connections (to different hosts).